docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+  @nn.compact
+  def __call__(self, x):
+    scale = self.param("scale", x.shape[-1:], nn.initializers.ones)
+    x *= scale
+    a = nn.Dense(self.hidden_size)(x)
+    h = nn.relu(a)
+    return nn.Dense(self.out_size)(h)
+```
+
+
+Many of the standard linen layers like `nn.Dense` use this type of shape inference already to avoid the need to specify input shapes (like the number of input features to a Dense layer).
+
+### Compact control flow
+
+The order in which you define submodules determines the name of a submodule if none is provided explicitly (using the name= construction argument). Because the name determines how parameters are mapped to submodules, you must be careful about mixing control flow with auto-generated names. Using control flow can change the order or remove certain submodules altogether. This is useful in the case where certain submodules should only exist depending on some construction argument. However, when control flow depends on the input arguments to the Module you should be careful. For example the following Module will break:
+
+
+```python
+class WrongModule(nn.Module):
+  @nn.compact
+  def __call__(self, x, mode):
+    if mode == "encode"
+      return nn.Dense(8)(x)
+    elif mode == "decode":
+      return nn.Dense(4)(x)
+```
+
+
+This problem can be solved in various ways: Provide explicit names, create the modules in setup, or move the constructor out of the control flow. The latter is done as follows:
+
+
+```python
+class CorrectModule(nn.Module):
+  @nn.compact
+  def __call__(self, x, mode):
+    encoder = nn.Dense(8)
+    decoder = nn.Dense(4)
+    if mode == "encode"
+      return encoder(x)
+    elif mode == "decode":
+      return decoder(x)
+```
+
+
+In the above example the construction order is fixed. After construction the submodules can be used in an arbitrary order.
+
+> Note \
+compact modules show a strong resemblance to [React hooks](https://reactjs.org/docs/hooks-custom.html)
+
+
+## Top-level Modules
+
+When a Module instance is created at the top-level it will be in an "unbound" state, meaning it has no variables attached. By top-level we mean it is not constructed as a sub-Module inside another Module class. Apart from calling `init` and `apply` there is not much you can do with an unbound Module. Note also that `setup` is not called on unbound Modules so you can really only access the construction arguments. See the Future work section for how this might change in the future.
+
+### Why are top-level modules always unbound?
+
+When we call `apply`, a copy of the top-level Module is created which will actually hold the variables and PRNG sequences. This stateful, "bound", clone only exists while we are executing the apply method. The reason for this is that if you create a stateful object and destroy it before the apply function returns, the `apply` function itself behaves like a pure function (pure means if you put the same arguments in, it will return the same outputs).
```

Comment:
This paragraph is not really clear to me. It seems to me that a function is pure when its outputs only depend on its inputs, regardless of whether you create stateful objects (and - or not - destroy them). So what's the motivation for the cloning exactly?
(An example might help)

Reply:
adding some clarification

Reply:
... but the function would still be pure if it returned the stateful object, no? if true, then the requirement of wanting to have a pure function would not necessitate you to destroy the bound module inside the function ...

Reply:
Here you get towards the problem of referential transparency. JAX expects a PyTree (freely cloneable because it doesn't have identity) object as output to a function. Cloning a stateful object is itself a side effect because the clone has it's own state. This is also why using dicts and lists with as Pytrees can lead to weird behavior (because they are actually stateful). For example if you have a function like this:

```python
def add_metrics(metrics, loss):
  metrics["loss"] = loss
  return metrics
```
The behavior will change if you add @jit to this function because the metrics dict will no longer be shared.
I try to avoid using referential transparency though as people find it very confusing. So I think it's a good to encourage pure function with immutable inputs/outputs. And anything else is "outside warranty" :)